semantics-3.0/Semantics.hs

```diff
@@ -1,34 +1,38 @@
-{-# LANGUAGE NamedFieldPuns #-}
 module Semantics where
 
-import           Data.List       (foldl')
 import           Data.Map.Strict (Map)
 import qualified Data.Map.Strict as Map
-import           Data.Set        (Set)
-import qualified Data.Set        as Set
 
-type SlotNumber = Integer
-type SlotInterval = (SlotNumber, SlotNumber)
+newtype Slot = Slot { getSlot :: Integer } deriving (Eq,Ord)
+
+instance Show Slot where
+    show (Slot n) = "(Slot " ++ show n ++ ")"
+
+newtype Ada = Lovelace { getLovelace :: Integer } deriving (Eq,Ord)
+
+instance Show Ada where
+    show (Lovelace n) = "(Lovelace " ++ show n ++ ")"
+
 type PubKey = Integer
 type Party = PubKey
 type NumChoice = Integer
 type NumAccount = Integer
-type Timeout = SlotNumber
-type Money = Integer
+type Timeout = Slot
+type Money = Ada
 type ChosenNum = Integer
 
 data AccountId = AccountId NumAccount Party
```

Comment:
what does `NumAccount` mean?

Reply:
and why not a record, this would avoid `accountOwner` function and the type aliases

Reply:
You could make the `show` instance be like `AccountId 1 2`

Reply:
I guess you can still use it like it is not a record... But then we either have to:
- Implement record syntax in the contract parser.
- Have a double semantics (it works like a "data" for whoever writes contracts and like a "record" internally").

Reply:
good points, agreed

Reply:
I realise that I think we are using a haskell parser on both the front end and back end so perhaps the parsing is not an issue. I also think that we could see what it looks like for the user to use it as a record. Perhaps we can try it on the front end and see.